### PR TITLE
GPU/CUDA Helpers 

### PR DESCRIPTION
**Pull request by samueltootle**
This PR consists of codegen helpers such as the `GPU_Kernel` class to provide a stencil for building GPU kernels and their correct function calls.

Additionally, CUDA utilities are included that are used for the transfer of data to/from host, populating `__constant__` memory (with, e.g. params), and pinned memory (de)allocation, and reductions.

Finally, a new header is added for double precision cuda intrinsics.  This is in the main helpers directory in the same manner as simd, but can be moved if desired.

### CODE REVIEW

File: nrpy/helpers/cuda_intrinsics.h [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+#define ConstCUDA(a) ((a))
+#define AbsCUDA(a) (fabs(a))
+#define AddCUDA(a, b) __dadd_rn((a), (b))
+#define SubCUDA(a, b) __dsub_rn((a), (b))
+#define MulCUDA(a, b) __dmul_rn((a), (b))
+#define DivCUDA(a, b) __ddiv_rn((a), (b))
+
+// Fused Multiply-Add/Subtract Operations (Scalar)
+#define FusedMulAddCUDA(a, b, c) __fma_rn((a), (b), (c))
+#define FusedMulSubCUDA(a, b, c) FusedMulAddCUDA((a), (b), MulCUDA((FDPart1_NegativeOne_), c))
```
Comment (zachetienne) [inline nrpy/helpers/cuda_intrinsics.h:17]: FDPart1_NegativeOne_ -> -1.0

File: nrpy/helpers/cuda_intrinsics.h [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+#define NegFusedMulAddCUDA(a, b, c) SubCUDA((c), MulCUDA((a), (b)))
+#define NegFusedMulSubCUDA(a, b, c) MulCUDA((FDPart1_NegativeOne_),(FusedMulAddCUDA((a), (b), (c))))
```
Comment (zachetienne) [inline nrpy/helpers/cuda_intrinsics.h:19]: FDPart1_NegativeOne_ -> -1.0

File: nrpy/helpers/cuda_intrinsics.h (hunk 1 above)
Comment (samueltootle) [reply nrpy/helpers/cuda_intrinsics.h]: Done

File: nrpy/helpers/cuda_intrinsics.h (hunk 2 above)
Comment (samueltootle) [reply nrpy/helpers/cuda_intrinsics.h]: Done

File: nrpy/tests/DOCTEST-gpu__cuda_utilities.json [hunk 3]
```diff
@@ -0,0 +1,12 @@
+{
```
Comment (zachetienne) [inline nrpy/tests/DOCTEST-gpu__cuda_utilities.json:1]: Please update to use `validate_strings()` (in nrpy/helpers/generic.py)

Comment (samueltootle) [reply nrpy/tests/DOCTEST-gpu__cuda_utilities.json]: All instances of using json have been replaced by `validate_strings()`

**Pull request merged**